### ISSUE TITLE
Automatically route with express

### ISSUE THREAD

**Issue by franciscop** [mxstbr/passport-magic-login#1]
The current documentation reads "Once you've got that, you'll then need to add a couple of routes to your Express server:"

```js
// Add the passport-magic-login strategy to Passport
passport.use(magicLogin)

// This is where we POST to from the frontend
app.post("/auth/magiclogin", magicLogin.send);

// This is what the user visits to confirm the login attempt and redirects them to the callbackUrl
app.get(magicLogin.confirmUrl, magicLogin.confirm);

// This is the standard Passport callbackUrl thing
app.get(magicLogin.callbackUrl, passport.authenticate("magiclogin"));
```

But we can instead make it a bit simpler to the user like this:

```js
app.use(magicLogin(passport));
```

Then inside magicLogin, we register the `passport.use()` and the different routes. This way the user doesn't need to do all of this scaffolding by themselves and the only work needed is:

```js
import MagicLoginStrategy from "passport-magic-login"

const magicLogin = new MagicLoginStrategy({
  // all the config
})

app.use(magicLogin(passport));
```

This is IMHO a much cleaner API.

No feature request is complete without the cons, so the main issue I see of this is the lack of transparency on what is going on internally. But this is _probably_ okay since the configuration key names are self-explanatory + there's documentation pointing out at this.

**Comment by mxstbr** [comment]
I think we'll also need a configuration key for the `app.post("/auth/magiclogin")` path. I wonder if we should just have a single configuration key for the root path or three separate ones?

```JS
const magicLogin = new MagicLoginStrategy({
  path: `/auth/magiclogin` // Will create /auth/magiclogin, /auth/magiclogin/confirm and /auth/magiclogin/callback
})

// vs.

const magicLogin = new MagicLoginStrategy({
  rootUrl: `/auth/magiclogin`,
  confirmUrl: `/auth/magiclogin/confirm`,
  callbackUrl: `/auth/magiclogin/callback`,
})
```

The former offers less control to the user, but at the same time I feel like almost no users would ever care about that level of control?

**Comment by franciscop** [comment]
It could even be simplified to this:

```js
import magicLogin from "passport-magic-login";

// A plain object with the config
const config = {};

app.use(magicLogin(config, passport));
```

**Comment by franciscop** [comment]
> ```js
> const magicLogin = new MagicLoginStrategy({
>   path: `/auth/magiclogin`
>   // Will create /${path}, /${path}/confirm and /${path}/callback routes
> })
> 
> // vs.
> 
> const magicLogin = new MagicLoginStrategy({
>   rootUrl: `/auth/magiclogin`,
>   confirmUrl: `/auth/magiclogin/confirm`,
>   callbackUrl: `/auth/magiclogin/callback`,
> })
> ```

This is exactly where I struggle, to pick either of those in this situation. Both seem quite good, first one is simpler but more magic-ish and second one is more explicit. I have a slight preference for the second one, but I think this is mostly an opinion-based thing at this level.

**Comment by rozenmd** [comment]
Hate to interject, but you could start with one:
```js
const magicLogin = new MagicLoginStrategy({
  path: `/auth/magiclogin`
  // Will create /${path}, /${path}/confirm and /${path}/callback routes
})
```
Then provide the other option:
```js
const magicLogin = new MagicLoginStrategy({
  rootUrl: `/auth/magiclogin`,
  confirmUrl: `/auth/magiclogin/confirm`,
  callbackUrl: `/auth/magiclogin/callback`,
})
```
If people care enough to ask for it?

(I don't think I've come across a situation where people wanted to control the callback urls when it comes to OAuth, but you never know)

**Comment by mxstbr** [comment]
I was also thinking that! What if we did the following:

```JS
const config = {
  path: `/auth/magiclogin`
  // Will create expand internally
}

// If the user wants more control, they can do the expansion manually:

const config = {
  path: {
    root: `/auth/magiclogin`,
    confirm: `/auth/magiclogin/confirm`,
    callback: `/auth/magiclogin/callback`,
  }
}
```

Then they can have both the control but also the convenience!

**Comment by franciscop** [comment]
So is this the API you like? :blush: 

```js
import magicLogin from "passport-magic-login";

const config = {
  path: `/auth/magiclogin`
  // Will create expand internally

  // ... other config
};

app.use(magicLogin(config, passport));
```

I can (probably?) make a PR with that

**Comment by franciscop** [comment]
Oh there's another, again more express-like alternative!

```js
import magicLogin from "passport-magic-login";

const config = {
  // ... other config
};

app.use(`/auth/magiclogin`, magicLogin(config, passport));
```

**Comment by mxstbr** [comment]
I prefer the former API (with the `path` configuration)! If you want to open a PR for that that'd be excellent ✨ Thank you!

**Comment by franciscop** [comment]
FYI this will need for either `express` to be a `peerDependency` to import the Router or do a bunch of `ifs` internally, I'll have a look at the code first to see what I think it's best but if you have a preference please let me know.

**Comment by mxstbr** [comment]
Adding express as a devDependency & peerDependency would be perfect!

**Comment by maraisr** [comment]
I disagree with this @mxstbr, granted I have no grounds.... But this would mean this lib would only make sense for express users.

Why not instead open up secondary entry point making `express` an peerDep (marked as optional, meta).

and do something like this:

```
import enhanceForExpress from 'passport-magic-login/express';
import PassportMagicLink from 'passport-magic-login';

const config = {
  path: `/auth/magiclogin`
  // ... other config
};

app.use(enhanceForExpress(magicLogin(passport), config));
```

the idea being, the lib itself operates purely with the `http` module, but allows for wiring up some things using these enhancers. This isnt to say that this package will now need to expose express, polka, koa—no, just the ones you care about.

**Comment by mxstbr** [comment]
I actually kinda like that! I wonder if we can just build a core module that works with plain HTTP and then add "enhancers" on top, while keeping the usage as simple as this:

```JS
// Plain HTTP usage
import MagicLoginStrategy from "passport-magic-login"

const magicLogin = new MagicLoginStrategy(...);

magicLogin.send(req: IncomingMessage, res: ServerResponseÇ)
magicLogin.confirm(req: IncomingMessage, res: ServerResponseÇ)
magicLogin.authenticate(req: IncomingMessage, res: ServerResponseÇ)
```

```JS
// Express usage
import magicLogin from "passport-magic-login/express"

app.use(magicLogin(...))
```

**Comment by franciscop** [comment]
Did some digging earlier on; seems like passport's API is specially designed to integrate neatly with Express/Connect [[1]](http://www.passportjs.org/docs/configure/#middleware) [[2]](https://stackoverflow.com/a/28543996/938236), so while _not_ having an express dependency could reach more people, does it make sense with the added complexity? It _is_ possible to use passport without express and there are hooks for it, but again [reading the docs on how to use passport](http://www.passportjs.org/docs/configure/#middleware) this feel like it was really not the original intention since they include a bunch of *required* express middleware:

```js
var session = require("express-session"),
    bodyParser = require("body-parser");

app.use(express.static("public"));
app.use(session({ secret: "cats" }));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(passport.initialize());
app.use(passport.session());
```

So this is the balance right, whether to target express or target a more general approach. I think both have their value, and after finding out about this I think targetting express directly is probably a better idea, but again I'll leave that decision to you!

> I actually kinda like that! I wonder if we can just build a core module that works with plain HTTP

Seeing how most of the workarounds work, it's probably possible to make it in a way that it works everywhere with a single import. This would be useful for those compatibility layers people seem to be doing to use passport with non-express, and also work with express as expected:

```js
// VERY simplified sample code, would need to properly parse the url+path, etc
module.exports = (opts) => {
  const callbackPath = "/auth/magiclogin/callback";

  return (req, res) => {
    if (req.method === "POST" && req.url === callbackPath) {
      // Do the logic for the callback
    }
    ...
  };
};
```

Usage on express would be as shown above in my snippet, and on other systems just like `magicLogin(opts)(req, res);`:

```js
import magicLogin from "passport-magic-login";

const config = {  /* ... */ };

// Express:
app.use(magicLogin(config, passport));

// No Express:
magicLogin(config, passport)(req, res);
```

My two takeaways are:
- Passport integration for Express is quite deep, does anyone here have experience using it without express? I think this might be quite rare seeing the docs of passport.
- Implementation with Express is a lot more natural and safe than doing a bunch of `ifs` to check for method, path, etc.

**Comment by mxstbr** [comment]
I use passport outside of Express with Next.js, but they have very similar `req/res` objects 😬 If it's simple to make it work without Express let's do that, if it's super tedious I'd rather we make it nice to use by default!

**Comment by JackCA** [comment]
I'm also interested in a framework agnostic approach because I'm using Koa and it doesn't have the same API as express -- but poking around at the code, there isn't much code that actually relies on express's api

**Comment by niklasgrewe** [comment]
@mxstbr Hi, thanks a lot for this strategy 👍 Is it possible to use it without Express? I am currently using MongoDB Realm as my backend system, similar to Firebase. I wonder if it is possible to use Passportjs with Serverless Functions? 🤔

**Comment by franciscop** [comment]
So unfortunately I tried doing this Issue but it was a bit more complex to setup than I thought and couldn't get it working. I see the API also has changed and now it's more intuitive than initially anyway, so please feel free to close it @mxstbr since it seems I won't be able to pull this out right now 😓

If you keep it open I might retry in the future when I have more time/energy to do the PR proper justice 😄

**Comment by mxstbr** [comment]
All good! Will keep this open in case anybody else wants to tackle this / you end up having time to do it!
